tests/e2e-kubernetes/testsuites/credentials.go

```diff
@@ -223,287 +239,311 @@ func (t *s3CSICredentialsTestSuite) DefineTests(driver storageframework.TestDriv
-			// TODO: How this should fail?
-		})
+					_, restoreDriverSA := overrideServiceAccountRole(ctx, f, driverSA, *driverRole.Arn)
+					deferCleanup(restoreDriverSA)
 
-		It("should not use csi driver's service account tokens", func(ctx context.Context) {
-			driverSA := getCSIDriverServiceAccount(ctx, f)
-			restoreDriverSA := overrideServiceAccountRole(ctx, f, driverSA, iamRoleS3FullAccess)
-			defer restoreDriverSA(ctx)
+					// Trigger recreation of CSI driver pods to use the new IAM role
+					killCSIDriverPods(ctx, f)
 
-			sa, deleteSA := createServiceAccount(ctx, f, "s3-csi-e2e-sa", annotateServiceAccountWithRole(iamRoleS3NoAccess))
-			defer deleteSA(ctx)
+					pod, _ := createPodWithServiceAccountAndPolicy(ctx, iamPolicyS3ReadOnlyAccess, true)
+					expectReadOnly(pod)
+				})
 
-			resource := storageframework.CreateVolumeResource(ctx, driver, l.config, pattern, t.GetTestSuiteInfo().SupportedSizeRange)
-			l.resources = append(l.resources, resource)
+				It("should not mix different pod's service account tokens even they are using same volume", func(ctx context.Context) {
+					vol := createVolumeResourceWithMountOptions(enablePodLevelIdentity(ctx), l.config, pattern, []string{"allow-delete"})
+					deferCleanup(vol.CleanupResource)
 
-			pod, err := createPodWithServiceAccount(ctx, f.ClientSet, f.Namespace.Name, []*v1.PersistentVolumeClaim{resource.Pvc}, sa.Name)
-			framework.ExpectNoError(err)
-			defer func() {
-				framework.ExpectNoError(e2epod.DeletePodWithWait(ctx, f.ClientSet, pod))
-			}()
+					saFullAccess := createServiceAccountWithPolicy(ctx, iamPolicyS3FullAccess)
+					saReadOnlyAccess := createServiceAccountWithPolicy(ctx, iamPolicyS3ReadOnlyAccess)
 
-			expectListToFail(pod)
-			expectWriteToFail(pod)
-		})
+					podFullAccess, err := createPodWithServiceAccount(ctx, f.ClientSet, f.Namespace.Name, []*v1.PersistentVolumeClaim{vol.Pvc}, saFullAccess.Name)
+					framework.ExpectNoError(err)
+					deferCleanup(func(ctx context.Context) error { return e2epod.DeletePodWithWait(ctx, f.ClientSet, podFullAccess) })
 
-		It("should not use mix different pod's service account tokens", func(ctx context.Context) {
-			saFullAccess, deleteSAFullAccess := createServiceAccount(ctx, f, "s3-csi-e2e-sa", annotateServiceAccountWithRole(iamRoleS3FullAccess))
-			defer deleteSAFullAccess(ctx)
+					podReadOnlyAccess, err := createPodWithServiceAccount(ctx, f.ClientSet, f.Namespace.Name, []*v1.PersistentVolumeClaim{vol.Pvc}, saReadOnlyAccess.Name)
+					framework.ExpectNoError(err)
+					deferCleanup(func(ctx context.Context) error { return e2epod.DeletePodWithWait(ctx, f.ClientSet, podReadOnlyAccess) })
 
-			saReadOnlyAccess, deleteSAReadOnlyAccess := createServiceAccount(ctx, f, "s3-csi-e2e-read-only-sa", annotateServiceAccountWithRole(iamRoleS3ReadOnlyAccess))
-			defer deleteSAReadOnlyAccess(ctx)
+					expectReadOnly(podReadOnlyAccess)
+					expectFullAccess(podFullAccess)
```

Comment:
Can we add another check that writing from full access pod is readable from read only pod? (And not writable)

Reply:
Added via https://github.com/awslabs/mountpoint-s3-csi-driver/pull/235/commits/b579efbb1656113be1745ee1cb09b7cd2025a0c3